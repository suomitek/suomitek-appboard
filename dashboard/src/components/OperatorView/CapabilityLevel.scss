.capabilityLevelIcon {
  position: relative;
  bottom: -0.4em;
  margin-right: 10px;
}
