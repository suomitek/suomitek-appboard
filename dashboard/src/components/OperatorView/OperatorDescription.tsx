import * as React from "react";
import ReactMarkdown from "react-markdown";

import HeadingRenderer from "../ChartView/HeadingRenderer";
import LinkRenderer from "../ChartView/LinkRenderer";

interface IOperatorDescriptionProps {
  description: string;
}

class OperatorDescription extends React.Component<IOperatorDescriptionProps> {
  public render() {
    return (
      <div className="ChartReadme">
        <ReactMarkdown
          source={this.props.description}
          renderers={{
            heading: HeadingRenderer,
            link: LinkRenderer,
          }}
          skipHtml={true}
        />
      </div>
    );
  }
}

export default OperatorDescription;
