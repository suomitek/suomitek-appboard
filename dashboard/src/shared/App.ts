import * as url from "shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { IAppOverview, IChartVersion } from "./types";

export const KUBEOPS_ROOT_URL = "api/tiller-deploy/v1";

export class App {
  public static async create(
    cluster: string,
    namespace: string,
    releaseName: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = url.kubeops.releases.list(cluster, namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    cluster: string,
    namespace: string,
    releaseName: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(
    cluster: string,
    namespace: string,
    releaseName: string,
    revision: number,
  ) {
    const endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(
    cluster: string,
    namespace: string,
    releaseName: string,
    purge: boolean,
  ) {
    let endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    if (purge) {
      endpoint += "?purge=true";
    }
    const { data } = await axiosWithAuth.delete(endpoint);
    return data;
  }

  public static async listApps(cluster: string, namespace?: string, allStatuses?: boolean) {
    let endpoint = namespace
      ? url.kubeops.releases.list(cluster, namespace)
      : url.kubeops.releases.listAll(cluster);
    if (allStatuses) {
      endpoint += "?statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(endpoint);
    return data.data;
  }

  public static async getRelease(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      url.kubeops.releases.get(cluster, namespace, name),
    );
    return data.data;
  }
}
