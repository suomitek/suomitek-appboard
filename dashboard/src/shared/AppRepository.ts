import { axiosWithAuth } from "./AxiosInstance";
import { APIBase } from "./Kube";
import { ICreateAppRepositoryResponse } from "./types";
import * as url from "./url";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(namespace));
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(namespace, name));
    return data;
  }

  public static async resync(name: string, namespace: string) {
    const repo = await AppRepository.get(name, namespace);
    repo.spec.resyncRequests = repo.spec.resyncRequests || 0;
    repo.spec.resyncRequests++;
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(namespace, name), repo);
    return data;
  }

  public static async update(
    name: string,
    namespace: string,
    repoURL: string,
    authHeader: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
  ) {
    const { data } = await axiosWithAuth.put<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.update(namespace, name),
      {
        appRepository: { name, repoURL, authHeader, customCA, syncJobPodTemplate, registrySecrets },
      },
    );
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(
      url.backend.apprepositories.delete(name, namespace),
    );
    return data;
  }

  // create uses the kubeapps backend API
  // TODO(mnelson) Update other endpoints to similarly use the backend API, removing the need
  // for direct k8s api access (for this resource, at least).
  public static async create(
    name: string,
    namespace: string,
    repoURL: string,
    authHeader: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
  ) {
    const { data } = await axiosWithAuth.post<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.create(namespace),
      {
        appRepository: { name, repoURL, authHeader, customCA, syncJobPodTemplate, registrySecrets },
      },
    );
    return data;
  }

  public static async validate(repoURL: string, authHeader: string, customCA: string) {
    const { data } = await axiosWithAuth.post<any>(url.backend.apprepositories.validate(), {
      appRepository: { repoURL, authHeader, customCA },
    });
    return data;
  }

  private static APIBase: string = APIBase;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apis/kubeapps.com/v1alpha1`;
  private static getSelfLink(namespace: string, name?: string): string {
    return `${AppRepository.APIEndpoint}/namespaces/${namespace}/apprepositories${
      name ? `/${name}` : ""
    }`;
  }
}
