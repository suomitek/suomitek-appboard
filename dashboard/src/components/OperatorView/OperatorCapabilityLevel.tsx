import * as React from "react";
import { CheckCircle, XCircle } from "react-feather";

import "./CapabilityLevel.css";

interface IOperatorCapabilitiesProps {
  level: string;
}

export const BASIC_INSTALL = "Basic Install";
export const SEAMLESS_UPGRADES = "Seamless Upgrades";
export const FULL_LIFECYCLE = "Full Lifecycle";
export const DEEP_INSIGHTS = "Deep Insights";
export const AUTO_PILOT = "Auto Pilot";

const levels = {
  [BASIC_INSTALL]: 1,
  [SEAMLESS_UPGRADES]: 2,
  [FULL_LIFECYCLE]: 3,
  [DEEP_INSIGHTS]: 4,
  [AUTO_PILOT]: 5,
};

class CapabiliyLevel extends React.Component<IOperatorCapabilitiesProps> {
  public render() {
    const { level } = this.props;
    const levelInt = levels[level];
    return Object.keys(levels).map(key => {
      return (
        <li key={key}>
          {levels[key] <= levelInt ? (
            <CheckCircle className="capabilityLevelIcon" stroke="#1598CB" />
          ) : (
            <XCircle className="capabilityLevelIcon" stroke="#C7C9C8" />
          )}
          <span style={{ color: levels[key] <= levelInt ? "inherit" : "#C7C9C8" }}>{key}</span>
        </li>
      );
    });
  }
}

export default CapabiliyLevel;
