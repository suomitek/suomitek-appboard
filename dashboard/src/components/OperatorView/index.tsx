import OperatorView from "./OperatorView";

export default OperatorView;
