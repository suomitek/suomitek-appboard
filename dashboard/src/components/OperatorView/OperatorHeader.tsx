import { RouterAction } from "connected-react-router";
import * as React from "react";
import { Link } from "react-router-dom";
import { app } from "../../shared/url";

interface IOperatorHeaderProps {
  id: string;
  icon?: string;
  hideButton?: boolean;
  disableButton?: boolean;
  description: string;
  cluster: string;
  namespace: string;
  version: string;
  provider: string;
  namespaced: boolean;
  push: (location: string) => RouterAction;
}

class OperatorHeader extends React.Component<IOperatorHeaderProps> {
  public state = {
    modalIsOpen: false,
  };

  public render() {
    const {
      id,
      icon,
      description,
      cluster,
      namespace,
      version,
      provider,
      hideButton,
      disableButton,
    } = this.props;
    return (
      <header>
        <div className="ChartView__heading margin-normal row">
          <div className="col-1 ChartHeader__icon">
            <div className="ChartIcon">
              <img className="ChartIcon__img" src={icon} alt="icon" />
            </div>
          </div>
          <div className="col-9">
            <div className="title margin-l-small">
              <h1 className="margin-t-reset">{id}</h1>
              <h5 className="subtitle margin-b-normal">
                {/* TODO(andresmgot): Filter by provider */}
                <span>{version} - Provided by </span>
                <Link to={app.operators.list(cluster, namespace)}>{provider}</Link>
              </h5>
              <h5 className="subtitle margin-b-reset">{description}</h5>
            </div>
          </div>
          {!hideButton && (
            <div className="col-2 ChartHeader__button">
              <button
                className="button button-primary button-accent"
                onClick={this.redirect}
                disabled={disableButton}
              >
                {disableButton ? "Deployed" : "Deploy"}
              </button>
            </div>
          )}
        </div>
        <hr />
      </header>
    );
  }

  public redirect = () => {
    this.props.push(app.operators.new(this.props.cluster, this.props.namespace, this.props.id));
  };
}

export default OperatorHeader;
