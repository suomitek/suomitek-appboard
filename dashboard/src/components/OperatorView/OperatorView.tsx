import * as React from "react";

import OperatorNotSupported from "components/OperatorList/OperatorsNotSupported";
import { RouterAction } from "connected-react-router";
import { Operators } from "../../shared/Operators";
import { IClusterServiceVersion, IPackageManifest } from "../../shared/types";
import { api } from "../../shared/url";
import { ErrorSelector } from "../ErrorAlert";
import UnexpectedErrorPage from "../ErrorAlert/UnexpectedErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import CapabiliyLevel from "./OperatorCapabilityLevel";
import OperatorDescription from "./OperatorDescription";
import OperatorHeader from "./OperatorHeader";

interface IOperatorViewProps {
  operatorName: string;
  operator?: IPackageManifest;
  getOperator: (namespace: string, name: string) => Promise<void>;
  isFetching: boolean;
  cluster: string;
  namespace: string;
  error?: Error;
  push: (location: string) => RouterAction;
  getCSV: (namespace: string, name: string) => Promise<IClusterServiceVersion | undefined>;
  csv?: IClusterServiceVersion;
}

class OperatorView extends React.Component<IOperatorViewProps> {
  public componentDidMount() {
    const { operatorName, namespace, getOperator } = this.props;
    getOperator(namespace, operatorName);
  }

  public componentDidUpdate(prevProps: IOperatorViewProps) {
    const { namespace, getOperator, getCSV } = this.props;
    if (prevProps.operator !== this.props.operator && this.props.operator) {
      const defaultChannel = Operators.getDefaultChannel(this.props.operator);
      if (defaultChannel) {
        getCSV(namespace, defaultChannel.currentCSV);
      }
    }
    if (prevProps.namespace !== this.props.namespace) {
      getOperator(this.props.namespace, this.props.operatorName);
    }
  }

  public render() {
    const { isFetching, cluster, namespace, operatorName, operator, error, push, csv } = this.props;
    if (cluster !== "default") {
      return <OperatorNotSupported namespace={namespace} />;
    }
    if (error) {
      return <ErrorSelector error={error} resource={`Operator ${operatorName}`} />;
    }
    if (isFetching || !operator) {
      return <LoadingWrapper />;
    }
    const channel = Operators.getDefaultChannel(operator);
    if (!channel) {
      return (
        <UnexpectedErrorPage
          text={`Operator ${operatorName} doesn't define a valid channel. This is needed to extract required info.`}
        />
      );
    }
    const { currentCSVDesc } = channel;
    const namespaced = currentCSVDesc.installModes.find(m => m.type === "AllNamespaces");
    return (
      <section className="ChartView padding-b-big">
        <OperatorHeader
          id={operator.metadata.name}
          description={currentCSVDesc.displayName}
          icon={api.operators.operatorIcon(this.props.namespace, operator.metadata.name)}
          version={currentCSVDesc.version}
          cluster={cluster}
          namespace={namespace}
          provider={operator.status.provider.name}
          namespaced={!namespaced?.supported}
          push={push}
          disableButton={!!csv}
        />
        <main>
          <div className="container container-fluid">
            <div className="row">
              <div className="col-9 ChartView__readme-container">
                <OperatorDescription description={currentCSVDesc.description} />
              </div>
              <div className="col-3 ChartView__sidebar-container">
                <aside className="ChartViewSidebar bg-light margin-v-big padding-h-normal padding-b-normal">
                  <div className="ChartViewSidebar__section">
                    <h2>Capability Level</h2>
                    <div>
                      <ul className="remove-style padding-l-reset margin-b-reset">
                        <li>
                          <CapabiliyLevel level={currentCSVDesc.annotations.capabilities} />
                        </li>
                      </ul>
                    </div>
                  </div>
                  <div className="ChartViewSidebar__section">
                    <h2>Repository</h2>
                    <div className="margin-l-big">
                      <span>
                        <a
                          href={currentCSVDesc.annotations.repository}
                          target="_blank"
                          rel="noopener noreferrer"
                        >
                          {currentCSVDesc.annotations.repository}
                        </a>
                      </span>
                    </div>
                  </div>
                  <div className="ChartViewSidebar__section">
                    <h2>Provider</h2>
                    <div className="margin-l-big">
                      <span>{operator.status.provider.name}</span>
                    </div>
                  </div>
                  <div className="ChartViewSidebar__section">
                    <h2>Container Image</h2>
                    <div className="margin-l-big">
                      <span>{currentCSVDesc.annotations.containerImage}</span>
                    </div>
                  </div>
                  <div className="ChartViewSidebar__section">
                    <h2>Created At</h2>
                    <div className="margin-l-big">
                      <span>{currentCSVDesc.annotations.createdAt}</span>
                    </div>
                  </div>
                </aside>
              </div>
            </div>
          </div>
        </main>
      </section>
    );
  }
}

export default OperatorView;
