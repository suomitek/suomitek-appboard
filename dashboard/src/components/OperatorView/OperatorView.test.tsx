import OperatorNotSupported from "components/OperatorList/OperatorsNotSupported";
import { shallow } from "enzyme";
import * as React from "react";
import { NotFoundError } from "../../shared/types";
import UnexpectedErrorPage from "../ErrorAlert/UnexpectedErrorAlert";
import OperatorDescription from "./OperatorDescription";
import OperatorHeader from "./OperatorHeader";
import OperatorView from "./OperatorView";

const defaultProps = {
  operatorName: "foo",
  getOperator: jest.fn(),
  isFetching: false,
  cluster: "default",
  namespace: "kubeapps",
  push: jest.fn(),
  getCSV: jest.fn(),
};

const defaultOperator = {
  metadata: {
    name: "foo",
    namespace: "kubeapps",
  },
  status: {
    provider: {
      name: "Kubeapps",
    },
    defaultChannel: "beta",
    channels: [
      {
        name: "beta",
        currentCSV: "foo.1.0.0",
        currentCSVDesc: {
          displayName: "Foo",
          version: "1.0.0",
          description: "this is a testing operator",
          annotations: {
            capabilities: "Basic Install",
            repository: "github.com/kubeapps/kubeapps",
            containerImage: "kubeapps/kubeapps",
            createdAt: "one day",
          },
          installModes: [],
        },
      },
    ],
  },
} as any;

it("displays an alert if rendered for an additional cluster", () => {
  const props = { ...defaultProps, cluster: "other-cluster" };
  const wrapper = shallow(<OperatorView {...props} />);
  expect(wrapper.find(OperatorNotSupported)).toExist();
});

it("calls getOperator when mounting the component", () => {
  const getOperator = jest.fn();
  shallow(<OperatorView {...defaultProps} getOperator={getOperator} />);
  expect(getOperator).toHaveBeenCalledWith(defaultProps.namespace, defaultProps.operatorName);
});

it("tries to get the CSV for the current operator", () => {
  const getCSV = jest.fn();
  const wrapper = shallow(<OperatorView {...defaultProps} getCSV={getCSV} />);
  wrapper.setProps({ operator: defaultOperator });

  expect(getCSV).toHaveBeenCalledWith(
    defaultOperator.metadata.namespace,
    defaultOperator.status.channels[0].currentCSV,
  );
});

it("re-request the operator when changing the namespace", () => {
  const getOperator = jest.fn();
  const wrapper = shallow(<OperatorView {...defaultProps} getOperator={getOperator} />);
  wrapper.setProps({ namespace: "other" });
  expect(getOperator).toHaveBeenCalledTimes(2);
});

it("shows an error if it exists", () => {
  const wrapper = shallow(<OperatorView {...defaultProps} error={new NotFoundError()} />);
  expect(wrapper.html()).toContain("Operator foo not found");
});

it("shows an error if the operator doesn't have any channel defined", () => {
  const operator = {
    status: {
      channels: [],
    },
  };
  const wrapper = shallow(<OperatorView {...defaultProps} operator={operator as any} />);
  expect(
    wrapper
      .find(UnexpectedErrorPage)
      .dive()
      .text(),
  ).toContain(
    "Operator foo doesn't define a valid channel. This is needed to extract required info",
  );
});

it("renders a full OperatorView", () => {
  const wrapper = shallow(<OperatorView {...defaultProps} operator={defaultOperator} />);
  expect(wrapper).toMatchSnapshot();
});

it("selects the default channel", () => {
  const operator = {
    ...defaultOperator,
    status: {
      ...defaultOperator.status,
      channels: [{ name: "alpha" }, defaultOperator.status.channels[0]],
    },
  };
  const wrapper = shallow(<OperatorView {...defaultProps} operator={operator as any} />);
  expect(wrapper.find(OperatorDescription).prop("description")).toEqual(
    "this is a testing operator",
  );
});

it("disables the Header deploy button if the CSV already exists", () => {
  const wrapper = shallow(
    <OperatorView {...defaultProps} operator={defaultOperator} csv={{} as any} />,
  );
  const header = wrapper.find(OperatorHeader);
  expect(header.prop("disableButton")).toBe(true);
});
