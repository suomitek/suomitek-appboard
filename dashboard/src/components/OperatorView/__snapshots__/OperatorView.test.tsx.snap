// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a full OperatorView 1`] = `
<section
  className="ChartView padding-b-big"
>
  <OperatorHeader
    cluster="default"
    description="Foo"
    disableButton={false}
    icon="api/v1/namespaces/kubeapps/operator/foo/logo"
    id="foo"
    namespace="kubeapps"
    namespaced={true}
    provider="Kubeapps"
    push={[MockFunction]}
    version="1.0.0"
  />
  <main>
    <div
      className="container container-fluid"
    >
      <div
        className="row"
      >
        <div
          className="col-9 ChartView__readme-container"
        >
          <OperatorDescription
            description="this is a testing operator"
          />
        </div>
        <div
          className="col-3 ChartView__sidebar-container"
        >
          <aside
            className="ChartViewSidebar bg-light margin-v-big padding-h-normal padding-b-normal"
          >
            <div
              className="ChartViewSidebar__section"
            >
              <h2>
                Capability Level
              </h2>
              <div>
                <ul
                  className="remove-style padding-l-reset margin-b-reset"
                >
                  <li>
                    <CapabiliyLevel
                      level="Basic Install"
                    />
                  </li>
                </ul>
              </div>
            </div>
            <div
              className="ChartViewSidebar__section"
            >
              <h2>
                Repository
              </h2>
              <div
                className="margin-l-big"
              >
                <span>
                  <a
                    href="github.com/kubeapps/kubeapps"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    github.com/kubeapps/kubeapps
                  </a>
                </span>
              </div>
            </div>
            <div
              className="ChartViewSidebar__section"
            >
              <h2>
                Provider
              </h2>
              <div
                className="margin-l-big"
              >
                <span>
                  Kubeapps
                </span>
              </div>
            </div>
            <div
              className="ChartViewSidebar__section"
            >
              <h2>
                Container Image
              </h2>
              <div
                className="margin-l-big"
              >
                <span>
                  kubeapps/kubeapps
                </span>
              </div>
            </div>
            <div
              className="ChartViewSidebar__section"
            >
              <h2>
                Created At
              </h2>
              <div
                className="margin-l-big"
              >
                <span>
                  one day
                </span>
              </div>
            </div>
          </aside>
        </div>
      </div>
    </div>
  </main>
</section>
`;
